## CLASSIFICATION TECHNIQUE

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.datasets import load_breast_cancer
from sklearn.feature_selection import SelectKBest, f_classif, mutual_info_classif, chi2
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.metrics import (accuracy_score, precision_score, recall_score, 
                           f1_score, roc_auc_score, confusion_matrix, 
                           classification_report, roc_curve)
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

In [3]:
# LOADING AND PREPROCESSING

In [4]:
data=load_breast_cancer()
x=pd.DataFrame(data.data, columns=data.feature_names)
y=pd.Series(data.target)
data

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [5]:
x.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
dtype: int64

In [6]:
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)
x_scaled_df = pd.DataFrame(x_scaled, columns=x.columns)


In [7]:
print("Feature Mean After Scaling:", x_scaled_df.mean().round(2).tolist()[:5])  # Checking the first 5 features
print("Feature Std Dev After Scaling:", x_scaled_df.std().round(2).tolist()[:5])

Feature Mean After Scaling: [-0.0, -0.0, -0.0, -0.0, 0.0]
Feature Std Dev After Scaling: [1.0, 1.0, 1.0, 1.0, 1.0]


-------------------

In [9]:
# CLASSIFICATION ALGORITHM IMPLEMENTATION 

In [10]:
# Split the dataset into training and testing sets
print("\nSplitting data...")

x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.2, random_state=42)


Splitting data...


In [11]:
#Initialize classifiers
classifiers = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(n_estimators=100, random_state=42),
    "SVM": SVC(kernel="linear", random_state=42),
    "k-NN": KNeighborsClassifier(n_neighbors=5)
}

In [18]:
for name, model in classifiers.items():
    model.fit(x_train,y_train)
    y_pred=model.predict(x_test)
    acc=accuracy_score(y_test,y_pred)
    results[name]=acc
    print(f"Model:{name}")
    print("\nClassification Report:")
    print(classification_report(y_test,y_pred))
    print("\nConfusion Matrix:")
    print(confusion_matrix(y_test, y_pred))
    print("-" * 50)

Model:Logistic Regression

Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114


Confusion Matrix:
[[41  2]
 [ 1 70]]
--------------------------------------------------
Model:Decision Tree

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.96      0.96      0.96        71

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114


Confusion Matrix:
[[40  3]
 [ 3 68]]
--------------------------------------------------
Model:Random Forest

Classification Report:
              precision    rec

In [12]:
results={}
for name, clf in classifiers.items():
    clf.fit(x_train, y_train)  # Train the model
    y_pred = clf.predict(x_test)  # Predict on the test set
    accuracy = accuracy_score(y_test, y_pred)  # Evaluate accuracy
    results[name] = accuracy

results

{'Logistic Regression': 0.9736842105263158,
 'Decision Tree': 0.9473684210526315,
 'Random Forest': 0.9649122807017544,
 'SVM': 0.956140350877193,
 'k-NN': 0.9473684210526315}

1- Logistic Regression achieved the highest accuracy (97.37%), likely due to the dataset being relatively well-suited to linear separation after feature scaling.

2- Random Forest performed comparably (96.49%), benefiting from its ability to handle feature interactions and reduce overfitting.

3- SVM (95.61%) effectively handled the high-dimensional data, showcasing its strength in separating classes with a linear kernel.

4- Decision Tree and k-NN had slightly lower accuracies (94.74%), with k-NN being sensitive to the choice of k and the scaled data distribution.

----------------------

In [ ]:
# MODEL COMPARISON

In [20]:
best_model = max(results, key=results.get)
worst_model = min(results, key=results.get)
print("\nPerformance Summary:")
for name, acc in results.items():
    print(f"{name}: Accuracy = {acc:.2f}")

print(f"\nBest Performing Model: {best_model} with Accuracy = {results[best_model]:.2f}")
print(f"Worst Performing Model: {worst_model} with Accuracy = {results[worst_model]:.2f}")


Performance Summary:
Logistic Regression: Accuracy = 0.97
Decision Tree: Accuracy = 0.95
Random Forest: Accuracy = 0.96
SVM: Accuracy = 0.96
k-NN: Accuracy = 0.95

Best Performing Model: Logistic Regression with Accuracy = 0.97
Worst Performing Model: Decision Tree with Accuracy = 0.95


-----------------------------------------------------------END-----------------------------------------------------------------